app/org/sagebionetworks/bridge/services/ConsentServiceImpl.java

```diff
@@ -168,19 +175,74 @@ public boolean hasUserSignedMostRecentConsent(StudyIdentifier studyIdentifier, U
     }
 
     @Override
-    public void withdrawConsent(Study study, User user) {
+    public void withdrawConsent(Study study, User user, Withdrawal withdrawal) {
         checkNotNull(study, Validate.CANNOT_BE_NULL, "study");
         checkNotNull(user, Validate.CANNOT_BE_NULL, "user");
-        if (userConsentDao.withdrawConsent(user.getHealthCode(), study)) {
-            decrementStudyEnrollment(study);
-            Account account = accountDao.getAccount(study, user.getEmail());
-            account.setConsentSignature(null);
+        checkNotNull(user, Validate.CANNOT_BE_NULL, "withdrawal");
+        
+        String externalId = optionsService.getOption(user.getHealthCode(), ParticipantOption.EXTERNAL_IDENTIFIER);
+        
+        Account account = accountDao.getAccount(study, user.getEmail());
+        signatureToHistory(account);
+        accountDao.updateAccount(study, account);
+
+        try {
+            userConsentDao.withdrawConsent(user.getHealthCode(), study, withdrawal.getWithdrewOn());    
+        } catch(Exception e) {
+            // Could not record the consent, compensate and rethrow the exception
+            signatureFromHistory(account);
             accountDao.updateAccount(study, account);
-            user.setConsent(false);
+            throw e;
         }
-        optionsService.deleteAllParticipantOptions(user.getHealthCode());
```

Comment:
If the user withdraws, do we want to set their sharing options to no_sharing? This matters if they user submits data (either actively or passively), then they withdraw. If we don't set their sharing options to no_sharing, their data will be exported (which based on previous conversations, is not a good thing).